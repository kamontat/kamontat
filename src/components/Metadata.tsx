/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.com/docs/use-static-query/
 */

import React from "react"
import { Helmet } from "react-helmet"
import { useIntl } from "gatsby-plugin-intl"

import { BaseOptions } from "../layout/Base"
import { useTheme } from "@emotion/react"

export type Meta = JSX.IntrinsicElements["meta"]

export interface SEOOptions extends BaseOptions {
  description: string
  meta?: Meta[]
  /**
   * page name
   */
  pageName: string
  /**
   * website title from graphql
   */
  title?: string
}

function SEO({ title, pageName, description, meta }: SEOOptions): JSX.Element {
  const intl = useIntl()
  const theme = useTheme()
  const defaultMeta: Meta[] = [
    {
      name: "title",
      content: title,
    },
    {
      name: `description`,
      content: description,
    },
    {
      property: `og:title`,
      content: title,
    },
    {
      property: `og:description`,
      content: description,
    },
    {
      property: `og:type`,
      content: `website`,
    },
    {
      name: `twitter:card`,
      content: `summary`,
    },
    {
      name: `twitter:title`,
      content: title,
    },
    {
      name: `twitter:description`,
      content: description,
    },
  ]

  return (
    <Helmet
      htmlAttributes={{
        lang: intl.locale,
        theme: theme.name,
      }}
      title={pageName}
      titleTemplate={title ? `%s | ${title}` : `%s`}
      meta={defaultMeta.concat(meta ?? [])}
    />
  )
}

export default SEO
