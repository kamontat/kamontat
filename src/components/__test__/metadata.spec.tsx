import React from "react"
import { render, waitFor } from "@testing-library/react"
import Metadata, { Html, SEOOptions } from "../Metadata"
import { IntlProvider } from "gatsby-plugin-intl"

describe("Metadata.Html", () => {
  it("with default configuration", async () => {
    render(<Html pageName="page-name" />)
    await waitFor(() => {
      expect(document.title).toEqual("page-name")
    })
  })

  it("with title", async () => {
    render(<Html pageName="page-name" title="website-title" />)
    await waitFor(() => {
      expect(document.title).toEqual("page-name | website-title")
    })
  })

  it("with single custom metadata tag", async () => {
    render(<Html pageName="page-name" meta={[{ name: "name", content: "content" }]} />)

    let data: NodeListOf<HTMLElementTagNameMap["meta"]> | undefined = undefined
    await waitFor(() => {
      expect(document.querySelectorAll("meta").length).toEqual(1)

      data = document.querySelectorAll("meta")
    })

    expect(data).toMatchSnapshot()
  })

  it("with multiple custom metadata tags", async () => {
    render(
      <Html
        pageName="page-name"
        meta={[
          { name: "name-1", content: "content-1" },
          { name: "name-2", content: "content-2" },
        ]}
      />
    )

    let data: NodeListOf<HTMLElementTagNameMap["meta"]> | undefined = undefined
    await waitFor(() => {
      expect(document.querySelectorAll("meta").length).toEqual(2)

      data = document.querySelectorAll("meta")
    })

    expect(data).toMatchSnapshot()
  })
})

const wrapIntl = (opt: SEOOptions) => (
  <IntlProvider locale="en">
    <Metadata {...opt} />
  </IntlProvider>
)

describe("Metadata", () => {
  it("Render page name", async () => {
    render(wrapIntl({ pageName: "page-name", description: "test" }))

    await waitFor(() => {
      expect(document.title).toEqual("page-name")
    })
  })

  it("Render custom meta", async () => {
    render(wrapIntl({ pageName: "page-name", description: "test", meta: [{ name: "name-1", content: "content-1" }] }))

    let data: NodeListOf<HTMLElementTagNameMap["meta"]> | undefined = undefined
    await waitFor(() => {
      expect(document.querySelectorAll("meta").length).toEqual(9)

      data = document.querySelectorAll("meta")
    })

    expect(data).toMatchSnapshot()
  })
})
