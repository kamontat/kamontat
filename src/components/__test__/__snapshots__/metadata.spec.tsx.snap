// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata Render custom meta 1`] = `
NodeList [
  <meta
    content=""
    data-react-helmet="true"
    name="title"
  />,
  <meta
    content="test"
    data-react-helmet="true"
    name="description"
  />,
  <meta
    content=""
    data-react-helmet="true"
    property="og:title"
  />,
  <meta
    content="test"
    data-react-helmet="true"
    property="og:description"
  />,
  <meta
    content="website"
    data-react-helmet="true"
    property="og:type"
  />,
  <meta
    content="summary"
    data-react-helmet="true"
    name="twitter:card"
  />,
  <meta
    content=""
    data-react-helmet="true"
    name="twitter:title"
  />,
  <meta
    content="test"
    data-react-helmet="true"
    name="twitter:description"
  />,
  <meta
    content="content-1"
    data-react-helmet="true"
    name="name-1"
  />,
]
`;

exports[`Metadata.Html with multiple custom metadata tags 1`] = `
NodeList [
  <meta
    content="content-1"
    data-react-helmet="true"
    name="name-1"
  />,
  <meta
    content="content-2"
    data-react-helmet="true"
    name="name-2"
  />,
]
`;

exports[`Metadata.Html with single custom metadata tag 1`] = `
NodeList [
  <meta
    content="content"
    data-react-helmet="true"
    name="name"
  />,
]
`;
